fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,463 @@
+            cache
+        })?;
+
+        self.last_committed_finality.write(finality.clone())?;
+
+        Ok(())
+    }
+}
+
+impl DefaultFinalityProvider {
+    pub fn new(config: Config, committed_finality: IPCParentFinality) -> Self {
+        let height_data = SequentialKeyCache::new(config.block_interval);
+        // nonce should be cached with increment 1
+        let top_down_msgs = SequentialKeyCache::new(1);
+
+        Self {
+            config,
+            parent_view_data: ParentViewData {
+                height_data: TVar::new(height_data),
+                top_down_msgs: TVar::new(top_down_msgs),
+            },
+            last_committed_finality: TVar::new(committed_finality),
+        }
+    }
+
+    fn check_height(&self, proposal: &IPCParentFinality) -> StmDynResult<()> {
+        let latest_height = if let Some(h) = self.parent_view_data.latest_height()? {
+            h
+        } else {
+            return abort(Error::HeightNotReady);
+        };
+
+        if latest_height < proposal.height {
+            return abort(Error::ExceedingLatestHeight {
+                proposal: proposal.height,
+                parent: latest_height,
+            });
+        }
+
+        let last_committed_finality = self.last_committed_finality.read()?;
+        if proposal.height <= last_committed_finality.height {
+            return abort(Error::HeightAlreadyCommitted(proposal.height));
+        }
+
+        Ok(())
+    }
+
+    fn check_block_hash(&self, proposal: &IPCParentFinality) -> StmDynResult<()> {
+        if let Some(block_hash) = self.parent_view_data.block_hash(proposal.height)? {
+            if block_hash == proposal.block_hash {
+                return Ok(());
+            }
+            return abort(Error::BlockHashNotMatch {
+                proposal: proposal.block_hash.clone(),
+                parent: block_hash,
+                height: proposal.height,
+            });
+        }
+        abort(Error::BlockHashNotFound(proposal.height))
+    }
+
+    fn check_validator_set(&self, proposal: &IPCParentFinality) -> StmDynResult<()> {
+        if let Some(validator_set) = self.parent_view_data.validator_set(proposal.height)? {
+            if validator_set != proposal.validator_set {
+                return abort(Error::ValidatorSetNotMatch(proposal.height));
+            }
+            return Ok(());
+        }
+        abort(Error::ValidatorSetNotFound(proposal.height))
+    }
+
+    fn check_top_down_msgs(&self, proposal: &IPCParentFinality) -> StmDynResult<()> {
+        if proposal.top_down_msgs.is_empty() {
+            // top down msgs are empty, no need checks
+            return Ok(());
+        }
+
+        ensure_sequential_by_nonce(&proposal.top_down_msgs)?;
+
+        let proposal_min_nonce = proposal.top_down_msgs.first().unwrap().msg.nonce;
+
+        let last_committed_finality = self.last_committed_finality.read()?;
+        if last_committed_finality.top_down_msgs.is_empty() {
```

Comment:
What if no new messages where included in the previous proposal?

Reply:
we will skip this top down messages check.